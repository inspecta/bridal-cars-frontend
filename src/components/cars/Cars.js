import React, { useEffect, useState } from 'react';
import { Link, useLocation, useNavigate } from 'react-router-dom';
import { FaCaretLeft, FaCaretRight } from 'react-icons/fa';
import { useDispatch, useSelector } from 'react-redux';
import PropTypes from 'prop-types';
import { fetchAllCars } from '../../redux/features/carSlice';
import Car from './Car';

const Cars = ({ showButton }) => {
  const dispatch = useDispatch();
  const navigate = useNavigate();
  const location = useLocation();
  const getAllCars = useSelector((state) => state.cars);
  const message = location.state?.message;

  const [currentPage, setCurrentPage] = useState(1);
  const [totalPages, setTotalPages] = useState(1);
  const [displayedCars, setDisplayedCars] = useState([]);
  const [carsPerPage, setCarsPerPage] = useState(1);

  useEffect(() => {
    dispatch(fetchAllCars());
  }, [dispatch]);

  useEffect(() => {
    const startIndex = (currentPage - 1) * carsPerPage;
    const endIndex = startIndex + carsPerPage;
    const carsToDisplay = getAllCars.cars.slice(startIndex, endIndex);
    setDisplayedCars(carsToDisplay);
    setTotalPages(Math.ceil(getAllCars.cars.length / carsPerPage));
  }, [currentPage, getAllCars.cars, carsPerPage]);

  useEffect(() => {
    const handleResize = () => {
      let newCarsPerPage;

      // Number of cars to display depending on screen size
      if (window.innerWidth >= 851) {
        newCarsPerPage = 3;
      } else if (window.innerWidth >= 600 && window.innerWidth <= 850) {
        newCarsPerPage = 2;
      } else {
        newCarsPerPage = 1;
      }

      setCarsPerPage(newCarsPerPage);
      setCurrentPage(1);
    };

    // Add evenlistener to the screen size
    window.addEventListener('resize', handleResize);
    handleResize();
    return () => window.removeEventListener('resize', handleResize);
  }, []);

  const handleClick = (carObject) => {
    navigate('/car-details', {
      state: {
        cars: carObject,
      },
    });
  };

  const handlePrevClick = () => {
    if (currentPage > 1) {
      setCurrentPage(currentPage - 1);
    }
  };

  const handleNextClick = () => {
    if (currentPage < totalPages) {
      setCurrentPage(currentPage + 1);
    }
  };

  return (
    <div className="cars-container">
      <h1 className="cars-container-header">LATEST BRIDAL CARS</h1>
      <p className="cars-container-subheader">Choose your bridal car today!</p>
      <div className="dotten-line" />
      <div>
        {message && <p>{message}</p>}
      </div>
      <div className="car-list-wrapper">
        <button
          onClick={handlePrevClick}
          disabled={currentPage === 1}
          type="button"
          className="btnPrev"
        >
          <FaCaretLeft />
        </button>
        <div className="cars-list">
          {
            displayedCars.length > 0
              ? displayedCars.map((car) => (
                <Car
                  key={car.id}
                  car={car}
                  onClick={() => handleClick(car)}
                  onKeyDown={() => handleClick()}
                  showButton={showButton}
                />
              ))
              : <p>No cars available.</p>
          }
        </div>
        <button
          onClick={handleNextClick}
          disabled={currentPage === totalPages}
          type="button"
          className="btnNext"
        >
          <FaCaretRight />
        </button>
      </div>
      {!showButton && (
        <div className="add-car-btn">
          <Link to="/add-car">Add Car</Link>
        </div>
      )}
    </div>
  );
};

Cars.propTypes = {
  showButton: PropTypes.bool.isRequired,
};
export default Cars;
