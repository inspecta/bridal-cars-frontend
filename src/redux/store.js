import { configureStore, combineReducers } from '@reduxjs/toolkit';
import carsReducer from './features/carSlice';
import userReducer from './features/user';
import alertReducer from './features/alert';

const rootReducer = combineReducers({
  cars: carsReducer,
  user: userReducer,
  alert: alertReducer,
});

const store = configureStore({
  reducer: rootReducer,
});

export default store;
