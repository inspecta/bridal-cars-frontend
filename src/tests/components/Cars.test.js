import React from 'react';
import { render } from '@testing-library/react';
import '@testing-library/jest-dom';
import { BrowserRouter } from 'react-router-dom';
import { Provider } from 'react-redux';
import store from '../../redux/store';
import Cars from '../../components/cars/Cars';

test('renders cars component snapshot', () => {
  const { asFragment } = render(
    <Provider store={store}>
      <BrowserRouter>
        <Cars />
      </BrowserRouter>
    </Provider>,
  );
  expect(asFragment()).toMatchSnapshot();
});
