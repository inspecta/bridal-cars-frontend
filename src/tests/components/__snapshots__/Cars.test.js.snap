// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders cars component snapshot 1`] = `
<DocumentFragment>
  <div>
    <div
      class="cars-container"
    >
      <div>
        <h1
          class="cars-container-header"
        >
          LATEST BRIDAL CARS
        </h1>
        <p
          class="cars-container-subheader"
        >
          Choose your bridal car today!
        </p>
      </div>
      <div
        class="dotten-line"
      />
      <div />
      <div
        class="car-list-wrapper"
      >
        <button
          class="btnPrev"
          disabled=""
          type="button"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 192 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M192 127.338v257.324c0 17.818-21.543 26.741-34.142 14.142L29.196 270.142c-7.81-7.81-7.81-20.474 0-28.284l128.662-128.662c12.599-12.6 34.142-3.676 34.142 14.142z"
            />
          </svg>
        </button>
        <div
          class="cars-list"
        >
          <p>
            No cars available.
          </p>
        </div>
        <button
          class="btnNext"
          type="button"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 192 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M0 384.662V127.338c0-17.818 21.543-26.741 34.142-14.142l128.662 128.662c7.81 7.81 7.81 20.474 0 28.284L34.142 398.804C21.543 411.404 0 402.48 0 384.662z"
            />
          </svg>
        </button>
      </div>
    </div>
    <div
      class="add-car-btn"
    >
      <a
        href="/add-car"
      >
        Add Car
      </a>
    </div>
  </div>
</DocumentFragment>
`;
